### ISSUE TITLE
[prefer-template] Add "allow-single-concat" option

### ISSUE THREAD

**Issue by ulrichb** [eslint/eslint#12042]
**What rule do you want to change?**
prefer-template

**Does this change cause the rule to produce more or fewer warnings?**
fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option "allow-single-concat"

**Please provide some example code that this change will affect:**

```js
const str = someVal + "";            // valid (with enabled option)
const str = "Val: " + someVal + "";  // invalid
```

**What does the rule currently do for this code?**
Error

**What will the rule do after it's changed?**
No error for "single" concat

**Background info**
I'm coming from TSLint, which offers this option. See https://palantir.github.io/tslint/rules/prefer-template/.

**Are you willing to submit a pull request to implement this change?**
Yes

**Comment by kaicataldo** [comment]
Hey, thanks for the issue. Do you mind providing some context for why you feel this option would be useful? Is it just to allow for string conversion?

**Comment by ulrichb** [comment]
Hi @kaicataldo,

I've just temporarily enabled `prefer-template` and checked the violations.

_In my code base_ (in ~ 95% of the violations) it's indeed about string conversions (always the pattern `xyz + ""`).

... whereas in TSLint's rule it's about string concats in general.

So actually, I want a "allow-string-conversions" option, (a "allow-single-concat" option would be / was a workaround).

If you want to know why string concat for conversion: That's because I have another (TypeScipt powered) rule, which requires template string expressions to have a _primitive type_. This rule would then conflict with string conversions à la `${xyz}` if `xyz` is non-primitive.

**Comment by ulrichb** [comment]
What do you think about a `"allowStringConversions"` option (which suppresses errors in the form of `x + ""` and `"" + x`, see https://github.com/eslint/eslint/issues/12042#issuecomment-517039518)?

If you like it I could try to prepare a PR.

**Comment by kaicataldo** [comment]
Sorry for the delay in response. I see the value in this option. I don’t think this rule should be forcing a specific style of type conversion.

**Comment by kaicataldo** [comment]
Let’s see what the rest of the team has to say.

**Comment by phaux** [comment]
For conversions you can always use `String(x)`

**Comment by ljharb** [comment]
For conversions, you *must* use String(x) or a template literal - using + empty string will invoke valueOf.

Even if this option exists, please ensure it warns on concatting the empty string, which is always incorrect.

**Comment by phaux** [comment]
My use case for `allowSingleConcat` option would be every time I want to prepend some short string to a big expression like:
```js
const url = "/foo?" + new URLSearchParams({
  // some big object
}).toString()
```
I think it looks much better than a template literal.

At the same time I doubt it's worth it since you can always split that expression into separate consts.

**Comment by mdjermanovic** [comment]
> For conversions, you _must_ use String(x) or a template literal - using + empty string will invoke valueOf.

The same applies to + non-empty string, so a 'regular' `prefer-template` autofix from `x + "foo" ` to `` `${x}foo` `` can potentially change behavior?

**Comment by ljharb** [comment]
@mdjermanovic yes, if `x` is an object that implements both valueOf and toString (of which only `Date` does among the builtins).

**Comment by mdjermanovic** [comment]
It could even make sense to ignore `x + ""` by default (but that would be probably a breaking change), since it's the responsibility of `no-implicit-coercion`.

Now we have two rules reporting and trying to fix the exact same range:

```js
/* eslint prefer-template:error */
/* eslint no-implicit-coercion:error */

x + ""
```

[Demo link](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IHByZWZlci10ZW1wbGF0ZTplcnJvciAqL1xuLyogZXNsaW50IG5vLWltcGxpY2l0LWNvZXJjaW9uOmVycm9yICovXG5cbnggKyBcIlwiIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)

Regardless of the order, `prefer-template` seems to get processed first and always wins the race, so this code becomes `` `${x  }` ``, although `String(x)` looks like a better fix in this case?

On the other hand, the following code should be indeed fixed to a template literal:

```js
/* eslint prefer-template:error */
/* eslint no-implicit-coercion:error */

x + "" + y
```

but, interestingly, it becomes `String(x) + y`

**Comment by ljharb** [comment]
The template literal is a better fix, because it's syntactic.

**Comment by ljharb** [comment]
@kaicataldo reopen? :-D

**Comment by kaicataldo** [comment]
In this case, it doesn't look like there was enough interest from the team. If someone would like to advocate for a change related to this, please feel free to create a new proposal!

**Comment by iamareebjamal** [comment]
Just imagine writing 
```js
`${glob.blah(new Option({ rescue: true })).fetch()}-fix`
console.error(`Error while downloading stuff ${model}`)
```

instead of 

```js
glob.blah(new Option({ rescue: true })).fetch() + '-fix'
console.error('Error while downloading stuff ' + model)
```

So much fun
